File: src/byte/foundation/bootstrap/register_providers.py
from typing import TYPE_CHECKING

from byte.foundation.bootstrap.bootstrapper import Bootstrapper

if TYPE_CHECKING:
    from byte.foundation import Application


class RegisterProviders(Bootstrapper):
    """Register all service providers with the application."""

    _merge: list[type] = []

    @staticmethod
    def merge(providers: list[type]) -> None:
        """Merge provider classes into the provider configuration before registration."""
        # Remove duplicates while preserving order
        seen = set()
        unique = []
        for provider in providers:
            if provider is not None and provider not in seen:
                seen.add(provider)
                unique.append(provider)
        RegisterProviders._merge = unique

    def bootstrap(self, app: Application) -> None:
        """Register all service providers with the application."""
        # Get providers from application configuration
        providers = RegisterProviders._merge

        # Register each provider
        for provider_class in providers:
            app.register(provider_class)

File: node/src/foundation/bootstrap/register-providers.ts
import type { Application } from '../application.ts'
import type { Bootstrapper } from './bootstrapper.ts'
import type { Newable } from '../../support/concerns/bootable.ts'
import type { ServiceProvider } from '../../support/service-provider.ts'

export class RegisterProviders implements Bootstrapper {
  static _merge: Newable<ServiceProvider>[] = []

  static merge(providers: Newable<ServiceProvider>[]): void {
    RegisterProviders._merge = [...RegisterProviders._merge, ...providers]
  }

  bootstrap(app: Application): void {
    for (const P of RegisterProviders._merge) {
      app.register(P)
    }
  }
}
